File: 0567-permutation-in-string/0567-permutation-in-string.py
"""
Time = O(n)
Space = O(1)
"""
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2):
            return False
        
        match = 0
        s1_count = [0] * 26
        s2_count = [0] * 26
        
        for i in range(len(s1)):
            s1_count[ord(s1[i]) - ord("a")] += 1
            s2_count[ord(s2[i]) - ord("a")] += 1
        
        for i in range(len(s1_count)):
            if s1_count[i] == s2_count[i]:
                match += 1
        
        left = 0
        for right in range(len(s1), len(s2)):
            if match == 26:
                return True
            
            index = ord(s2[right]) - ord("a")
            s2_count[index] += 1
            if s1_count[index] == s2_count[index]:
                match += 1
            elif s1_count[index] + 1 == s2_count[index]:
                match -= 1
            
            index = ord(s2[left]) - ord("a")
            s2_count[index] -= 1
            if s1_count[index] == s2_count[index]:
                match += 1
            elif s1_count[index] - 1 == s2_count[index]:
                match -= 1
            
            left += 1
        
        return match == 26
            
                
                

File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        size_t m = s1.size(), n = s2.size();
        if (m > n) {
            return false;
        }
        // create counter for current window
        vector<int> counter(26, 0);
        for (int i = 0; i < m; i++) {
            counter[s1[i] - 'a']++;
            counter[s2[i] - 'a']--;
        }

        // count zeros, if all zeros means matched
        if (isMatched(counter)) {
            return true;
        }
        
        // slide the window to the right and check each time
        for (int i = m; i < n; i++) {
            counter[s2[i] - 'a']--;
            counter[s2[i - m] - 'a']++;
            if (isMatched(counter)) {
                return true;
            }
        }

        return false;
    }

private:
    bool isMatched(vector<int> &counter) {
        for (auto c: counter) {
            if (c != 0) {
                return false;
            }
        }

        return true;
    }
};